src/Rings/AbelianClosure.jl

```diff
@@ -843,6 +846,86 @@ function Oscar.order(a::QQAbElem)
   return o
 end
 
+
+###############################################################################
+#
+#   Embeddings of subfields of cyclotomic fields
+#   (works for proper subfields of cycl. fields only if these fields
+#   have been constructed as such)
+#
+
+# Construct the map from `F` to an abelian closure `K` such that `gen(F)`
+# is mapped to `x`.
+# If `F` is a cyclotomic field with conductor `N` then assume that `gen(F)`
+# is mapped to `QQAbElem(gen(F), N)`.
+# (Use that the powers of this element form a basis of the field.)
+function _embedding(F::AnticNumberField, K::QQAbField{AnticNumberField},
+                    x::QQAbElem{nf_elem})
+  R, = polynomial_ring(QQ, "x")
+  fl, n = Hecke.is_cyclotomic_type(F)
+  if fl
+    # This is cheaper.
+    f = function(x::nf_elem)
+      return QQAbElem(x, n)
+    end
+
+    finv = function(x::QQAbElem; check::Bool = false)
+      if n % conductor(x) == 0
+        return Hecke.force_coerce_cyclo(F, data(x))
+      elseif check
+        return
+      else
+        error("element has no preimage")
+      end
```

Comment:
What are the intended semantics of `check` here (and is this stated somewhere)? I am just surprised to see that `check == true` *disables* the error, and instead we return `nothing` (making the return value of this function somewhat type unstable)

Reply:
The point is that in the case that the given element has no preimage, the code is intended to be used in two different situations:
An exception shall be thrown if one has just asked for a preimage, but the function shall signal the nonexistence of a preimage if one has asked to check for the existence of a preimage; here this is done via a `nothing` return value.

I know that this looks dirty, but the problem is that the `MapFromFunc` objects have access only to their stored data (domain, codomain, and the two Julia functions). Note that the function for computing preimages needs some objects from the context of its creation, thus we cannot simply duplicate the code "outside the creation of the mapping".

(Would it be better to implement `has_preimage` via `try`/`catch` and `preimage`?)